### PR TITLE
HYC-1439: Emulated amd64 Docker on M1 processors (i.e. aarch64/arm64v8)

### PR DESCRIPTION
**Pull request by cazzerson**
This is an attempt at a universal Docker config for amd64/x86 and aarch64/arm64v8 (M1). It will run in an emulated environment on M1 platform. This includes some dependency updates that will need to be tested. This branch hasn't yet been tested on an x86 machine. The containers that I built on an M1 Mac have been pushed to the repository in the `docker-compose.yml` file so `mutagen-compose pull` from this branch should work.

### CODE REVIEW

**Comment by cazzerson** [comment]
I removed the gem and ruby updates to simplify the integration of this PR.

File: bin/yarn [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env ruby
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply bin/yarn:1]: Ah, weird! I will also revert all of these binstubs. I regenerated them early on to try to address some other issues I was running into. These are all stock rails binstubs.

File: docker-compose.yml [hunk 2]
```diff
@@ -67,7 +70,7 @@ services:
     ports:
       - "6379:6379"
   solr6:
-    image: solr:6-alpine
+    image: solr:6.6.6-alpine
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply docker-compose.yml:73]: https://www.youtube.com/watch?v=WxnN05vOuSM

File: config/environments/development.rb [hunk 3]
```diff
@@ -62,12 +68,14 @@
...
+  # config.file_watcher = ActiveSupport::EventedFileUpdateChecker
+  config.file_watcher = ActiveSupport::FileUpdateChecker
 
-  # Allow Vagrant IP address to display web console in development mode
+  # Allow Docker host IP address to display web console in development mode
   # NOTE: When we upgrade to Web Console 4.x this will change to
-  # config.web_console.permissions = ['10.0.2.2']
-  config.web_console.whitelisted_ips = ['10.0.2.2']
+  # config.web_console.permissions = ['172.20.0.1']
+  config.web_console.whitelisted_ips = ['172.20.0.1']
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply config/environments/development.rb:78]: Yes, good point.

File: build_log.txt [hunk 4]
```diff
@@ -0,0 +1,1190 @@
+➜  hy-c git:(m_one_docker) ✗ docker compose build web
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply build_log.txt:1]: Yep!

File: Dockerfile [hunk 5]
```diff
@@ -14,39 +14,49 @@ VOLUME [ "/sys/fs/cgroup" ]
...
-COPY . /hyrax
+ENV PATH "/fits:$PATH"
+COPY docker/fits.xml /fits/fits-1.5.5/xml/fits.xml
 
+# Install gems
+COPY Gemfile* /hyrax/
 WORKDIR /hyrax
-RUN scl enable rh-ruby26 -- bundle install
+
+#RUN scl enable devtoolset-8 rh-ruby26 -- gem update --system \
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply Dockerfile:52]: Nope!

File: Dockerfile [hunk 6]
```diff
@@ -14,39 +14,49 @@ VOLUME [ "/sys/fs/cgroup" ]
...
-WORKDIR /fits
-ADD https://github.com/harvard-lts/fits/releases/download/1.5.5/fits-1.5.5.zip /fits/
-RUN unzip fits-1.5.5.zip -d /fits && \
-rm -rf fits-1.5.5.zip
-ENV PATH "/fits:$PATH"
+# devtoolset-8 installed due to newer mini_racer requirement of newer g++
+# Also install ChromeDriver
+# TODO: are we using httpd?
+#RUN yum -y update  && yum -y install httpd && yum clean all && systemctl enable httpd.service \
+RUN yum -y install centos-release-scl-rh centos-release-scl \
```
[the remark this answers is not in the document]
Comment (cazzerson) [reply Dockerfile:26]: Removed the commented line. The massive RUNs are there to significantly reduce the number of Docker layers. 

https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers

It's ugly and weird. Might be more elegant to transition to multi-stage builds, but this was the shorter path for now.

**Comment by cazzerson** [comment]
> I think this works with docker, but not on my vagrant vm due to complains about yarn. I was able to deploy to cdr-test using this branch, but I needed to update hyc-gems using the vagrant vm in order to do so. Max's current instructions on confluence involve using docker for this purpose: https://confluence.lib.unc.edu/display/RAT/Release+Hy-C+and+dependencies But my impression from this PR is that it wouldn't be usable for that purpose now?

This docker container should actually be usable for packaging hyc-gems, but I haven't tested this on our servers.

File: config/environments/development.rb [hunk 7]
```diff
@@ -12,6 +12,12 @@
   # Show full error reports.
   config.consider_all_requests_local = true
 
+  # Where to store cached assets
+
+  config.assets.configure do |env|
+    env.cache = ActiveSupport::Cache.lookup_store(:file_store, "/opt/hyrax/rails-cache/assets/#{Rails.env}/")
```
Comment (bbpennel) [inline config/environments/development.rb:18]: I think this line is now failing the build in the vagrant vm
```
Errno::EACCES: Permission denied @ dir_s_mkdir - /opt/hyrax
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache/file_store.rb:158:in `ensure_cache_path'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache/file_store.rb:83:in `write_entry'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache/strategy/local_cache.rb:150:in `write_entry'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache.rb:446:in `block in write'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache.rb:663:in `block in instrument'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/notifications.rb:170:in `instrument'
/home/vagrant/.gem/ruby/gems/activesupport-5.2.7.1/lib/active_support/cache.rb:663:in
...
```
I think in vagrant all the hyrax stuff goes into `/hyrax/`, although i don't see a `rails-cache` in there at the moment. We might need to need to use a variable here to specify that path. There are the existing `DATA_STORAGE` and `TEMP_STORAGE` paths, both of which contain cache data, maybe we should direct this to one of those?

Comment (cazzerson) [reply config/environments/development.rb]: I just pushed a commit that makes this path configurable. I had been trying to keep the Rails-specific cache separate from the Hyrax one, but that may not be necessary.

**Review by bbpennel — APPROVED** [review]

**Pull request merged**